src/neuroglancer/datasource/deepzoom/frontend.ts

```diff
@@ -0,0 +1,240 @@
+
+  getSources(volumeSourceOptions: VolumeSourceOptions) {
+    const {rank} = this;
+    const chunkDataSizes = [Uint32Array.of(this.info.tilesize, this.info.tilesize, 3)];
+    return transposeNestedArrays(this.info.levels.map((levelInfo, index, array) => {
+      const relativeScale = 1 << index;
+      const stride = rank + 1;
+      const chunkToMultiscaleTransform = new Float32Array(stride * stride);
+      chunkToMultiscaleTransform[chunkToMultiscaleTransform.length - 1] = 1;
+      const {upperBounds: baseUpperBound} =
+          this.info.modelSpace.boundingBoxes[0].box;
+      const upperClipBound = new Float32Array(rank);
+      for (let i = 0; i < 2; ++i) {
+        chunkToMultiscaleTransform[stride * i + i] = relativeScale;
+        upperClipBound[i] = baseUpperBound[i] / relativeScale;
+      }
+      chunkToMultiscaleTransform[stride * 2 + 2] = 1;
+      upperClipBound[2] = baseUpperBound[2];
+      return makeDefaultVolumeChunkSpecifications({
+               rank,
+               dataType: this.dataType,
+               chunkToMultiscaleTransform,
+               upperVoxelBound: Float32Array.of(levelInfo.width, levelInfo.height, 3),
+               volumeType: this.volumeType,
+               chunkDataSizes,
+               volumeSourceOptions,
+             })
+          .map((spec): SliceViewSingleResolutionSource<VolumeChunkSource> => ({
+                 chunkSource: this.chunkManager.getChunkSource(DeepzoomImageTileSource, {
+                   credentialsProvider: this.credentialsProvider,
+                   spec,
+                   parameters: {
+                     url: resolvePath(this.url, (array.length - 1 - index).toString()),
+                     encoding: this.info.encoding,
+                     format: this.info.format,
+                     overlap: this.info.overlap,
+                     tilesize: this.info.tilesize
+                   }
+                 }),
+                 chunkToMultiscaleTransform,
+                 upperClipBound,
+               }));
+    }));
+  }
+}
+
+interface DZIMetaData {
+  width: number;
+  height: number;
+  tilesize: number;
+  overlap: number;
+  format: string;
+}
+
+function getDZIMetadata(
+    chunkManager: ChunkManager, credentialsProvider: SpecialProtocolCredentialsProvider,
+    url: string): Promise<DZIMetaData> {
+  if (url.endsWith('.json') || url.includes('.json?')) {
+    /* http://openseadragon.github.io/examples/tilesource-dzi/
+     * JSON variant is a bit of a hack, it's not known how much it is in use for real.
+     * The actual reason for not implementing it right now is the lack of CORS-enabled
+     * test data.
+     */
+    throw new Error('DZI-JSON: OpenSeadragon hack not supported yet.');
+  }
+  return chunkManager.memoize.getUncounted(
+      {'type': 'deepzoom:metadata', url, credentialsProvider: getObjectId(credentialsProvider)},
+      async () => {
+        return await cancellableFetchSpecialOk(credentialsProvider, url, {}, responseText)
```

Comment:
You can simplify this to use javascript async functions rather than `Promise.then`.

```
const text = await cancellableFetchSpecialOk(credentialsProvider, url, {}, responseText);
const xml = ...
```

Reply:
Yes, of course, that's just stuck this way after the initial copy-paste. `return await` without a `try-catch` around is suspicious anyway, there's another one in the same file.
However I'm leaving for a week right now, so the changes will have to wait.

Reply:
Done in https://github.com/google/neuroglancer/pull/453/commits/a5737feb51489a7b612d8bee5a376612ef25b608